acceptance/helpers.rb

```diff
@@ -0,0 +1,230 @@
+require 'beaker-puppet'
+
+module Beaker
+  module DSL
+    # Host selectors by role
+    module Roles
+      # Select any hosts which have the agent role and are not the master.
+      # Beaker's `agents` selector selects all of the hosts that have the
+      # `agent` role, but some masters may have _both_ the agent and master
+      # roles. This `agents_only` selector will not include any agent hosts that
+      # have the `master` role.
+      #
+      # @return [Array<Beaker::Host>] A set of beaker hosts which have the `agent` role but not the `master` role
+      def agents_only
+        hosts_as(:agent).reject { |host| host['roles'].include?('master') }.to_a
+      end
+    end
+
+    # Helpers for testing the puppetlabs-puppet_agent module with Beaker
+    module Helpers
+      # Purging puppet-agent between the tests requires a helper script for some
+      # platforms (for example, on Windows). This directory holds those scripts.
+      SUPPORTING_FILES = File.expand_path('./files').freeze
+
+      # These platforms are only supported under Puppet Enterprise, and FOSS
+      # tests should be skipped on them.
+      PE_ONLY_PLATFORMS = %w[aix amazon sles solaris osx]
+
+      # Use `puppet module install` to install the puppet_agent module's
+      # dependencies on the target host, and then install the puppet_agent
+      # module itself. Requires that puppet is installed in advance.
+      #
+      # @param [Beaker::Host] host The target host
+      def install_modules_on(host)
+        on(host, puppet('module', 'install', 'puppetlabs-stdlib',     '--version', '5.1.0'), { acceptable_exit_codes: [0] })
+        on(host, puppet('module', 'install', 'puppetlabs-inifile',    '--version', '2.4.0'), { acceptable_exit_codes: [0] })
+        on(host, puppet('module', 'install', 'puppetlabs-apt',        '--version', '6.0.0'), { acceptable_exit_codes: [0] })
+
+        install_dev_puppet_module_on(host,
+                                     source: File.join(File.dirname(__FILE__), '..', ),
+                                     module_name: 'puppet_agent')
+      end
+
+      # Install puppet-agent on all agent nodes and connect them to the master.
+      # This is intended to prepare SUTs for tests that upgrade puppet-agent
+      # from some initial version.
+      #
+      # @param [String] initial_package_version_or_collection Either a version
+      #   of puppet-agent or the name of a puppet collection to install the agent from.
+      def prepare_upgrade_with(initial_package_version_or_collection)
```

Comment:
`prepare_upgrade_with` is too specific and should be something left to the tests instead of handled at the helper level. How about

```
setup_puppet_agent_on(hosts, package_version_or_collection, options = {})
```

and you can specify the master as an option so that the SSL cert-sign stuff can be conditionally triggered. This way, you ca re-use this helper in the master set-up stage without losing the clarity (since you're basically doing the same thing there).